DEV-569

shahbaz-bemeir opened the pull request:
… shipment emails

shahbaz-bemeir committed:
added admin configuration under Shipstream Sync for sending shipment emails

shahbaz-bemeir committed:
inventory_sync: subtracted Magento order items qty that are in processing state and not submitted to shipstream

shahbaz-bemeir committed:
Used payload data to create shipment in openmage for only shipped items

shahbaz-bemeir committed:
inventory_sync: added check for on hold orders

shahbaz-bemeir committed:
shipment:packed event data handling in magento shipment
- Managed packages with qty shipped in decimal portion
- Ship all remaining items in case order is completed on shipstream
- Foramted & added comments data in shipment

colinmollenhour commented on app/code/community/ShipStream/Sync/Model/Cron.php:
```diff
@@ -28,10 +28,21 @@ public function fullInventorySync($sleep = TRUE)
+                            if ( ! isset($target[$sku])) {
+                                continue;
+                            }
+                            $syncQty = $qty =  floor(floatval($qty));
```
Accidental assignment of `$qty`?

shahbaz-bemeir replied:
`$qty` is storing floor of qty in case as suggested on Clickup ticket `For inventory sync use floor operation on the quantities as there are cases where ShipStream may have a decimal quantity but the Magento product does not expect there to be decimals - such as "cases" of products that have been split at the warehouse but can only be ordered in whole quantities by the customer`

colinmollenhour commented on app/code/community/ShipStream/Sync/Model/Cron.php:
```diff
@@ -28,10 +28,21 @@ public function fullInventorySync($sleep = TRUE)
+                                $syncQty = floor(bcsub($qty, floatval($processingQty[$sku]['qty'], 4)));
+                            }
+                            $targetQty = floatval($target[$sku]['qty']);
+                            if ($syncQty === $targetQty) {
```
This may work but it is safer to use bccomp with scale 0:
```suggestion
                            if (bccomp($syncQty, $targetQty, 0) === 0) {
```

shahbaz-bemeir replied:
`bccomp` function is creating issues with `inventory_sync` method call. So for now I am not using this

colinmollenhour commented on app/code/community/ShipStream/Sync/Model/Cron.php:
```diff
@@ -90,7 +101,34 @@ protected function _getTargetInventory(array $skus)
+            ->from(['soi' => $resource->getTableName('sales/order_item')], $columns)
+            ->join(['so' => $resource->getTableName('sales/order')], 'so.entity_id = soi.order_id', [])
+            ->where('so.state NOT IN (?)',$orderStates)
+            ->where('so.status != (?)',"submitted")
+            ->where('so.hold_before_status != (?)',"submitted")
+            ->where('soi.sku IN (?)', $skus)
+            ->where('soi.product_type = (?)', 'simple')
```
I've never seen the right hand side wrapped in parentheses for = or !=, looks odd..

colinmollenhour commented on app/code/community/ShipStream/Sync/Model/Cron.php:
```diff
@@ -90,7 +101,34 @@ protected function _getTargetInventory(array $skus)
+        ];
+        $resource = Mage::getSingleton('core/resource');
+        $db = $resource->getConnection('core_write');
+        $columns = ['sku' => 'soi.sku', 'qty' => 'sum(soi.qty_ordered)'];
```
Would it be more accurate to use `sum(soi.qty_ordered - soi.qty_shipped - soi.qty_canceled)`?

colinmollenhour commented on app/code/community/ShipStream/Sync/Model/Order/Shipment/Api.php:
```diff
@@ -124,4 +131,104 @@ public function createWithTracking($orderIncrementId, $data)
+                    $itemShippedQty[$key] = bcadd($itemShippedQty[$key], $item['order_item_qty'],4);
+                }
+                else {
+                    $itemShippedQty[$key] = floatval($item['order_item_qty']);
```
Not necessary to use floatval when bcmath is used.
```suggestion
                    $itemShippedQty[$key] = $item['order_item_qty'];
```

colinmollenhour commented on app/code/community/ShipStream/Sync/Model/Order/Shipment/Api.php:
```diff
@@ -124,4 +131,104 @@ public function createWithTracking($orderIncrementId, $data)
+        if(function_exists('yaml_emit')) {
+            return yaml_emit($comments);
+        } else {
+            return json_encode($comments);
```
Make it more human-readable:

```suggestion
            return json_encode($comments, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE);
```

colinmollenhour commented on app/code/community/ShipStream/Sync/Model/Cron.php:
```diff
@@ -90,7 +101,34 @@ protected function _getTargetInventory(array $skus)
+     * @param array $sku
+     * @return array
+     */
+    protected function _getProcessingOrderItemsQty(array $skus): array
```
Please remove type declarations, even though 7.3 is EOL it may still be broadly in use.
```suggestion
    protected function _getProcessingOrderItemsQty(array $skus)
```

colinmollenhour commented on app/code/community/ShipStream/Sync/Model/Order/Shipment/Api.php:
```diff
@@ -124,4 +131,104 @@ public function createWithTracking($orderIncrementId, $data)
+     * @param $data
+     * @return array
+     */
+    protected function _getShippedItemsQty($order, $data): array
```
```suggestion
    protected function _getShippedItemsQty($order, $data)
```

colinmollenhour commented on app/code/community/ShipStream/Sync/Model/Order/Shipment/Api.php:
```diff
@@ -124,4 +131,104 @@ public function createWithTracking($orderIncrementId, $data)
+     * @param $data
+     * @return string
+     */
+    protected function _getCommentsData($order, $data): string
```
```suggestion
    protected function _getCommentsData($order, $data)
```

shahbaz-bemeir committed:
code refactoring to use natve methods and made functions type safe for php7.x

shahbaz-bemeir committed:
updated github workflow to create release on code push in main branch

shahbaz-bemeir committed:
code formatting and refactoring
- sync_inventory: Handled qty_refunded for processing orders on inventory sync
- shipment:pacled event: handling fractional BOM items

shahbaz-bemeir committed:
shipment:packed topic : get  order item reference Ids from shipment data to create shipment in magento

colinmollenhour commented on app/code/community/ShipStream/Sync/Model/Order/Shipment/Api.php:
```diff
@@ -124,4 +131,111 @@ public function createWithTracking($orderIncrementId, $data)
+            // Handling fractional BOM items shipped qty
+            $fraction = fmod($ordered_qty, 1);
+            $wholeNumber = intval($ordered_qty);
+            if ($fraction >= 0.88) {
```
Why 0.88? I think the floating point error should be very small, probably 0.9999 would be good enough to catch all floating point arithmetic errors but still allow for some edge cases like 29/30 of the total shipped.
```suggestion
            if ($fraction >= 0.9999) {
```

colinmollenhour committed:
Change rounding tolerance.

colinmollenhour committed:
Change cron config to do hourly sync instead of daily.
